IMDA-BTG/aiverify
# re-addChore/tests select dataset model

## Description

PR contains mostly unit tests updates for canvas and user input screen (select model , dataset and fill inputs).
Also contains minor improvement to when the `saved` message is displayed when save button is clicked on header.

- Most unit tests code changes are under `__tests__` and `__mock__` subfolders (also include jest files)
- Code changes under `ai-verify-portal/src/modules/project/header.tsx` are additions of lines on save button click handling. The existing `save` function is called and no change is made to that. Additional new codes are added after it to ensure that the `Project saved` message is displayed only after save is successful. Previously, it was just displayed without checking if save is successful. With this improvement, unit test can assert the APIs that's needed to be called when save is clicked.
- All other code changes that has additions of `toErrorWithMessage` simply improves the error message formatting (out put are normally just logged in browser console). No impact to functionality.
- There is 1-liner code change in `ai-verify-portal/src/modules/projectTemplate/canvas.tsx` where `new Date().getTime()` is changed to `Date.now()`. Both return the exact same thing. Changing to `Date.now()` allows us to stub and mock it in unit tests`
- Another addition in `ai-verify-portal/src/modules/projectTemplate/canvas.tsx` are `isNaN` handlings which only happens in unit tests jsdom, because there are no screen pixel level calculations or details available.

## Motivation and Context

Increase unit tests coverage

## Type of Change




 - chore: Routine tasks, maintenance, or tooling changes 








## How to Test

1. Under ai-verify-portal root folder, run `npm test`
2. On report canvas, ensure that widget drag & drop behaviours are normal
3. Click save icon on header and see that `Project Saved` message is displayed 



## Checklist

Please check all the boxes that apply to this pull request using "x":

- [ x]  I have tested the changes locally and verified that they work as expected.
- [x ]  I have added appropriate unit tests or functional tests for the changes made.
- [x ]  I have followed the project's coding conventions and style guidelines.
- [x ]  I have squashed or reorganized my commits into logical units.
- [x ]  I have performed a self-review of my own code.
- [x ]  I have read, understood and agree to the Developer Certificate of Origin below, which this project utilises.

## Screenshots (if applicable)

![image](https://github.com/IMDA-BTG/aiverify/assets/130328114/d1695283-e466-4009-8da2-f613b81ddd33)


## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]

<details>
  <summary>Developer Certificate of Origin</summary>

 ```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
 ```
</details>
